cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java

```diff
@@ -1101,6 +1159,45 @@ protected Future<Void> updateNodePoolStatuses(KafkaStatus kafkaStatus) {
         }
     }
 
+    /**
+     * Validate the Kafka version set in the Kafka custom resource, together with the metadata version and the configured
+     * inter.broker.protocol.version and log.message.format.version.
+     * They need to be all aligned and at least 3.4 to support ZooKeeper to KRaft migration.
+     *
+     * @param kafkaStatus Status of the Kafka custom resource where warnings about any issues with metadata state will be added
+     *
+     * @return Future which completes when validation is done
+     */
+    Future<Void> validateVersionsForKRaftMigration(KafkaStatus kafkaStatus) {
+        KafkaMetadataConfigurationState kafkaMetadataConfigState = kafkaMetadataStateManager.getMetadataConfigurationState();
+        if (kafkaMetadataConfigState.isPreMigration()) {
+            // validate 3.4 <= kafka.version && metadataVersion/IBP/LMF == kafka.version
+
+            MetadataVersion kafkaVersion = MetadataVersion.fromVersionString(kafka.getKafkaVersion().version());
+            // this should check that spec.kafka.version is >= 3.4
+            boolean isMigrationSupported = kafkaVersion.isMigrationSupported();
```

Comment:
This does not seem to make sense to me. You do not want to do migration together with an upgrade. So you have to expect that the Kafka version before starting migration will be at least one of the supported versions (e.g. 3.5 right now, 3.6 in the future etc.)

Reply:
Actually the minimum version is 3.4 not 3.5. This is what `isMigrationSupported` checks https://github.com/apache/kafka/blob/3.6.1/server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java#L309

Reply:
But nobody will start the migration from Kafka 3.4 because we will not support Kafka 3.4 at the point when the migration is supported.

Reply:
Or to put it differently, if someone tries to start migration with Kafka 3.4.0, you should tell them to first complete the upgrade. Because otherwise you are mixing up Kafka upgrade and migration into one process.

Reply:
That's a good point. I was leveraging the meaning of "is migration supported" from a Kafka perspective (which is 3.4) and not from a Strimzi perspective (which will be 3.6 when migration will be available).
I will change this check, by having 3.6 as minimum (which means not using the `isMigrationSupported` from Kafka anymore).

Reply:
I wonder if you are mixing two different checks:
* The cluster should likely use metadata version of at least 3.4
* The cluster should use a supported Kafka version (i.e. not 3.5 or 3.6 as one day it will run with a Strimzi version supporting only 3.8 and 4.0)

Reply:
You are right. The second one would mean updating the operator each time we get rid of a Kafka version so it doesn't make any sense.

Coming back to what the code is doing now. The migration is supported in Kafka only if the metadata version is at least 3.4. Thinking more, isn't it what we are more interested in? When the user updates the operator to the new version with migration support, it won't support 3.4 anymore so as we know the user will be already "forced" to upgrade to something like 3.6. The checks above will be valid and migration will be allowed.
I think that the check as it stands today is valid, but are you worried about trying to start a migration together with an upgrade? Is it really possible to happen? I am not sure about that.

Reply:
> When the user updates the operator to the new version with migration support, it won't support 3.4 anymore so as we know the user will be already "forced" to upgrade to something like 3.6. The checks above will be valid and migration will be allowed.

I'm not 100% sure it is worth the effort to implement it in code -> maybe docs are enough. But at least in theory, you might want to check if the upgrade is already complete.

> I think that the check as it stands today is valid, but are you worried about trying to start a migration together with an upgrade? Is it really possible to happen? I am not sure about that.

The check should pass I think -> but it does not seem to do anything useful.

Reply:
> The check should pass I think -> but it does not seem to do anything useful.

It should be always true, taking into account operator version supporting migration will have Kafka at least 3.6 (and blocking the usage of any older 3.4 for example). We could even delete this check, just adding doc. But it doesn't hurt that much. I would leave it for now, and see at the time of merging the PR.

Reply:
In theory, a user could set `.spec.kafka.metadataVersion: 3.3` on a Strimzi version supporting ZK-KRaft migration, so I think this check is needed.